### PR TITLE
Improve RootCommand executable name handling in NAOT app and library scenarios

### LINKED ISSUE #2812: IndexOutOfRangeException in RootCommand ctor
Within my [project](https://github.com/CasualPokePlayer/GSE), an exception ends up thrown in RootCommand's ctor.
```
System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at System.CommandLine.RootCommand.get_ExecutablePath() + 0x4c
   at System.CommandLine.RootCommand.get_ExecutableName() + 0x24
   at System.CommandLine.RootCommand..ctor(String) + 0x1c
   at GSE.CliArgs.Parse(String[]) + 0x6c
   at GSE.Entrypoint.Main(String[]) + 0x14
```
This occurs within the Android (or rather linux-bionic) version of this project, compiled under NativeAOT. Looking at [the ExecutablePath getter](https://github.com/dotnet/command-line-api/blob/fa1991f84bc8c384aa636a251398a40e56ee1702/src/System.CommandLine/RootCommand.cs#L77), this seems to be due to `Environment.GetCommandLineArgs()` returning an empty array (thus indexing it throws).

### PR DESCRIPTION
**Pull request by jonsequitur**
This pull request introduces several improvements and bug fixes to how `RootCommand` determines the executable name and path, especially in advanced hosting scenarios like NativeAOT native libraries. It adds comprehensive tests to ensure correct behavior, addresses edge cases with command-line parsing, and improves packaging for build targets.

**Executable Name/Path Resolution Improvements:**

* Refactored `RootCommand.ExecutableName` and `RootCommand.ExecutablePath` to more robustly determine the executable name and path, including fallback to an `AppContext` value injected by build targets when no command line arguments are available (e.g., in NativeAOT shared library scenarios) (`src/System.CommandLine/RootCommand.cs`).
* Added the `System.CommandLine.targets` file to the NuGet package to ensure the executable-name module initializer is available for all consumers, including test projects and native libraries (`src/System.CommandLine/System.CommandLine.csproj`).

**NativeAOT and Advanced Hosting Test Coverage:**

* Added a new test project (`NativeLibrary`) that builds as a NativeAOT shared library, exporting a native symbol to validate that `RootCommand.Name` falls back to the correct value when `Environment.GetCommandLineArgs()` is empty (`src/System.CommandLine.Tests/TestApps/NativeLibrary/NativeLibrary.csproj`, `src/System.CommandLine.Tests/TestApps/NativeLibrary/Program.cs`). [[1]](diffhunk://#diff-1727a8b69294450cf416e559d0f86a9c570f84d704e5fd4c897998c0c0f6fbcfR1-R32) [[2]](diffhunk://#diff-99a44b65ca1c4f998dd529aab8c87bdd982de886b702efc8e1b43eedad9400d8R1-R32)
* Introduced an end-to-end test in `CompilationTests.cs` to verify the fallback logic for `RootCommand.Name` in native library hosting scenarios, including platform-specific handling and dynamic invocation of the exported native symbol (`src/System.CommandLine.Tests/CompilationTests.cs`).

**Command-Line Parsing Bug Fixes and Tests:**

* Fixed a bug where option values resembling paths (e.g., `/p:Key=something/myapp`) could incorrectly be treated as root command names, and added targeted regression tests for these scenarios (`src/System.CommandLine/Parsing/StringExtensions.cs`, `src/System.CommandLine.Tests/ParserTests.RootCommandAndArg0.cs`). [[1]](diffhunk://#diff-ad5a560bcdb379aa20858352ddd00c5946e179ca13a033e4acce9d7b70cdb9b9R278-R284) [[2]](diffhunk://#diff-039e38295f8d4d110129383e61b761d301a7a09dd572c435b7db69d549df28f0L86-R219)
* Enhanced tests for root command and argument matching, including new assertions on error-free parsing and correct command resolution for various input patterns (`src/System.CommandLine.Tests/ParserTests.RootCommandAndArg0.cs`). [[1]](diffhunk://#diff-039e38295f8d4d110129383e61b761d301a7a09dd572c435b7db69d549df28f0R20-R42) [[2]](diffhunk://#diff-039e38295f8d4d110129383e61b761d301a7a09dd572c435b7db69d549df28f0R66-R88)

**General Test Improvements:**

* Added basic tests to ensure `RootCommand.ExecutablePath` and `RootCommand.ExecutableName` are never null or empty (`src/System.CommandLine.Tests/RootCommandTests.cs`).

**Build and Target Framework Updates:**

* Updated test and solution files to target .NET 10.0, ensuring compatibility with new features and NativeAOT scenarios (`System.CommandLine.v3.ncrunchsolution`).

These changes collectively improve the reliability and correctness of executable name detection across diverse hosting models, strengthen the test suite, and fix subtle parsing bugs.

### CODE REVIEW

File: src/System.CommandLine/build/System.CommandLine.targets [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+  <PropertyGroup>
+    <_SystemCommandLineExecutableName Condition="'$(_SystemCommandLineExecutableName)' == ''">$(AssemblyName)</_SystemCommandLineExecutableName>
+  </PropertyGroup>
+
+  <PropertyGroup>
+    <_SystemCommandLineGeneratedFile>$(IntermediateOutputPath)System.CommandLine.ExecutableName.g.cs</_SystemCommandLineGeneratedFile>
+  </PropertyGroup>
+
+  <Target Name="_GenerateSystemCommandLineExecutableName"
```
Comment (baronfel) [inline src/System.CommandLine/build/System.CommandLine.targets]: this needs to be conditioned on `Language` or else it'd break C#, VB, etc consumers.

File: src/System.CommandLine/build/System.CommandLine.targets [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+          BeforeTargets="CoreCompile"
+          Inputs="$(MSBuildAllProjects)"
```
Comment (baronfel) [inline src/System.CommandLine/build/System.CommandLine.targets]: IIRC MSBuildAllProjects as an input is an antipattern - if you care about properties as inputs, the pattern today you should use is

* have a target before your main target that writes the properties you care about into a file in the IntermeidateOutputPath with `WriteOnlyWhenDifferent='true'`
* have that file be the input to your main target
* make sure to add any files you generate to the `FileWrites` Item type in MSbuild so that `dotnet clean` clears them out

**Review by baronfel — CHANGES_REQUESTED** [review]
Logic looks fine but the MSBuild targets need some work to not break incrementality

File: src/System.CommandLine/build/System.CommandLine.targets [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    </PropertyGroup>
+
+    <!-- Escape semicolons so WriteLinesToFile does not treat them as MSBuild list
+         separators (which would strip them from the generated C# source). -->
+    <WriteLinesToFile File="$(_SystemCommandLineGeneratedFile)"
+                     Lines="$(_SystemCommandLineGeneratedCode.Replace(';', '%3B'))"
+                     Overwrite="true"
+                     WriteOnlyWhenDifferent="true" />
+
+  </Target>
```
Comment (baronfel) [inline src/System.CommandLine/build/System.CommandLine.targets]: Found a better way to set the data - you can add 

```xml
  <RuntimeHostConfigurationOption Include="System.CommandLine.ExecutableName" Value="$(_SystemCommandLineExecutableName)" />
```

and you get language-independence.  This works by putting the value in the runtimeconfig.json that the SDK emits and that the coreclr (and NAOT) loads up when starting.

Comment (baronfel) [reply src/System.CommandLine/build/System.CommandLine.targets]: This worked really well so I've pushed it up to minimize the work in the targets

**Review by baronfel — APPROVED** [review]

**Review by JeremyKuhne — APPROVED** [review]
Looks good to me. The SDK is following the same pattern to push root context.

**Review by baronfel — APPROVED** [review]

**Review by adamsitnik — APPROVED** [review]
LGTM, thank you @jonsequitur and @baronfel !

File: src/System.CommandLine.Tests/CompilationTests.cs [hunk 4]
```diff
@@ -42,6 +44,119 @@ public void App_referencing_system_commandline_can_be_compiled_ahead_of_time()
...
+            {
+                // On Windows the native library may still be locked after NativeLibrary.Free
+            }
+        }
+    }
+
+    private static string NativeLibraryFileName(string assemblyName) =>
+        OperatingSystem.IsWindows() ? $"{assemblyName}.dll"
+            : OperatingSystem.IsMacOS() ? $"{assemblyName}.dylib"
+            : $"{assemblyName}.so";
```
Comment (akoeplinger) [inline src/System.CommandLine.Tests/CompilationTests.cs:125]: > [Fix native library filename: NativeAOT doesn't add lib prefix](https://github.com/dotnet/command-line-api/pull/2820/commits/1e0af7a6a6f137bae32d749bb14a9d3a8097850d)

FYI this changed with .NET 11: https://github.com/dotnet/docs/issues/52324
